src/test/java/jenkins/plugins/git/GitToolChooserTest.java

```diff
@@ -0,0 +1,495 @@
+        GitTool jgitTool = new JGitTool(Collections.<ToolProperty<?>>emptyList());
+        GitTool jGitApacheTool = new JGitApacheTool(Collections.<ToolProperty<?>>emptyList());
+        jenkins.jenkins.getDescriptorByType(GitTool.DescriptorImpl.class).setInstallations(tool, jgitTool, jGitApacheTool);
+
+        buildAProject(sampleRepo, false);
+
+        List<TopLevelItem> list = jenkins.jenkins.getItems();
+
+        // Assuming no tool is installed and git is present in the machine
+        String gitExe = tool.getGitExe();
+
+        GitToolChooser sizeEstimator = new GitToolChooser(remote, list.get(0), "github", gitExe);
+        assertThat(sizeEstimator.getGitTool(), is("/usr/bin/git"));
+    }
+    /*
+    A test extension implemented to clone the behavior of a plugin extending the capability of providing the size of
+    repo from a remote URL of "Github".
+     */
+    @TestExtension
+    public static class TestExtensionGithub extends GitToolChooser.RepositorySizeAPI {
+
+        @Override
+        public boolean isApplicableTo(String remote, Item context, String credentialsId) {
+            return remote.contains("github");
+        }
+
+        @Override
+        public Long getSizeOfRepository(String remote, Item context, String credentialsId) {
+            // from remote, remove .git and https://github.com
+            return (long) 500;
+        }
+    }
+
+    /*
+    A test extension implemented to clone the behavior of a plugin extending the capability of providing the size of
+    repo from a remote URL of "GitLab".
+     */
+    @TestExtension
+    public static class TestExtensionGitlab extends GitToolChooser.RepositorySizeAPI {
+
+        @Override
+        public boolean isApplicableTo(String remote, Item context, String credentialsId) {
+            return remote.contains("gitlab");
+        }
+
+        @Override
+        public Long getSizeOfRepository(String remote, Item context, String credentialsId) {
+            // from remote, remove .git and https://github.com
+            return (long) 10000;
+        }
+    }
+
+    /*
+    A test extension implemented to clone the behavior of a plugin extending the capability of providing the size of
+    repo from a remote URL of "BitBucket".
+     */
+    @TestExtension
+    public static class TestExtensionBit extends GitToolChooser.RepositorySizeAPI {
+
+        @Override
+        public boolean isApplicableTo(String remote, Item context, String credentialsId) {
+            return remote.contains("bit");
+        }
+
+        @Override
+        public Long getSizeOfRepository(String remote, Item context, String credentialsId) throws IOException {
+            throw new IOException();
+        }
+    }
+
+
+    private void buildAProject(GitSampleRepoRule sampleRepo, boolean noCredentials) throws Exception {
+        WorkflowJob p = jenkins.jenkins.createProject(WorkflowJob.class, "p");
```

Comment:
Since the test has an option to be called without credentials, shouldn't the test without credentials skip passing credentials?
```suggestion
        WorkflowJob p = jenkins.jenkins.createProject(WorkflowJob.class, "p");
        String creds = noCredentials ? "" : "credentialsId: 'github', ";
```

Reply:
Since these credentials are used to validate a REST connection from the extension points and the `TestExtension` extensions do not perform an actual validation, I didn't add this. But rightly pointed out, this is should be added. 

I was wondering if I should improve the implemented extension tests with at least adding a scan on passed credentials to simulate that validation part of validating a passed down remote url.